html2docx.py

```diff
@@ -14,138 +15,106 @@ def __init__(self, title: str):
         super().__init__()
         self.doc = Document()
         self.doc.core_properties.title = title
-
-        self.p: Paragraph = None
-        self.runs: List[Run] = []
         self.list_style: List[str] = []
-        self.bold = 0
-        self.italic = 0
-        self.underline = 0
-        self.extra_data = ""
-
-    def feed(self, data: str) -> None:
-        super().feed(data)
-        self.finish()
+        self.href = ""
+        self._reset()
+
+    def _reset(self) -> None:
+        self.p: Optional[Paragraph] = None
+        self.r: Optional[Run] = None
+
+        # Formatting options
+        self.style = None
+        self.pre = False
+        self.attrs: List[List[str]] = []
+        self.collapse_space = True
+
+    def finish_p(self) -> None:
+        if self.r is not None:
+            self.r.text = self.r.text.rstrip()
+        self._reset()
+
+    def init_run(self, attrs: List[str]) -> None:
```

Comment:
`attrs` is always a list of one element (except for `a` tags, where it’s empty. Handled below).

The code is simpler as follows:

```diff
diff --git a/html2docx.py b/html2docx.py
index d885334..8110aa7 100644
--- a/html2docx.py
+++ b/html2docx.py
@@ -26,7 +26,7 @@ class HTML2Docx(HTMLParser):
         # Formatting options
         self.style = None
         self.pre = False
-        self.attrs: List[List[str]] = []
+        self.attrs: List[str] = []
         self.collapse_space = True
 
     def finish_p(self) -> None:
@@ -34,9 +34,9 @@ class HTML2Docx(HTMLParser):
             self.r.text = self.r.text.rstrip()
         self._reset()
 
-    def init_run(self, attrs: List[str]) -> None:
+    def init_run(self, attr: str) -> None:
         self.r = None
-        self.attrs.append(attrs)
+        self.attrs.append(attr)
 
     def finish_run(self) -> None:
         self.attrs = self.attrs[:-1]
@@ -48,9 +48,8 @@ class HTML2Docx(HTMLParser):
             self.p = self.doc.add_paragraph(style=style)
         if self.r is None:
             self.r = self.p.add_run()
-            for style_set in self.attrs:
-                for run_style in style_set:
-                    setattr(self.r.font, run_style, True)
+            for style in self.attrs:
+                setattr(self.r.font, style, True)
         self.r.add_text(data)
 
     def add_list_style(self, name: str) -> None:
@@ -63,14 +62,13 @@ class HTML2Docx(HTMLParser):
     def handle_starttag(self, tag: str, attrs: List[Tuple[str, Optional[str]]]) -> None:
         if tag == "a":
             self.href = str(dict(attrs).get("href", ""))
-            self.init_run([])
         elif tag in ["b", "strong"]:
-            self.init_run(["bold"])
+            self.init_run("bold")
         elif tag == "br":
             if self.r:
                 self.r.add_break()
         elif tag in ["em", "i"]:
-            self.init_run(["italic"])
+            self.init_run("italic")
         elif tag in ["h1", "h2", "h3", "h4", "h5", "h6"]:
             level = int(tag[-1])
             self.p = self.doc.add_heading(level=level)
@@ -79,7 +77,7 @@ class HTML2Docx(HTMLParser):
         elif tag == "pre":
             self.pre = True
         elif tag == "u":
-            self.init_run(["underline"])
+            self.init_run("underline")
         elif tag == "ul":
             self.add_list_style("List Bullet")
 
@@ -99,7 +97,7 @@ class HTML2Docx(HTMLParser):
             self.add_text(data)
 
     def handle_endtag(self, tag: str) -> None:
-        if tag in ["a", "b", "em", "i", "strong", "u"]:
+        if tag in ["b", "em", "i", "strong", "u"]:
             self.finish_run()
         elif tag in ["h1", "h2", "h3", "h4", "h5", "h6", "li", "ol", "p", "pre", "ul"]:
             self.finish_p()
```